flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/InsertOrUpdateJdbcExecutor.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.java.io.jdbc.executor;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.java.io.jdbc.JdbcStatementBuilder;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * {@link JdbcBatchStatementExecutor} that provides upsert semantics by updating row if it exists and inserting otherwise.
+ * Used in Table API.
+ */
+@Internal
+public final class InsertOrUpdateJdbcExecutor<R, K, V> implements JdbcBatchStatementExecutor<R> {
+
+	private final String existSQL;
+	private final String insertSQL;
+	private final String updateSQL;
+
+	private final JdbcStatementBuilder<K> existSetter;
+	private final JdbcStatementBuilder<V> insertSetter;
+	private final JdbcStatementBuilder<V> updateSetter;
+
+	private final Function<R, K> keyExtractor;
+	private final Function<R, V> valueMapper;
+
+	private transient PreparedStatement existStatement;
+	private transient PreparedStatement insertStatement;
+	private transient PreparedStatement updateStatement;
+	private transient Map<K, V> batch = new HashMap<>();
+
+	public InsertOrUpdateJdbcExecutor(
+		String existSQL,
+		String insertSQL,
+		String updateSQL,
+		JdbcStatementBuilder<K> existSetter,
+		JdbcStatementBuilder<V> insertSetter,
+		JdbcStatementBuilder<V> updateSetter,
+		Function<R, K> keyExtractor,
+		Function<R, V> valueExtractor) {
+		this.existSQL = existSQL;
```

Comment:
I think it makes sense to add null checks here so that the contracts of the class are explicit.